kernel/crates/kernel_physical_memory/src/lib.rs

```diff
@@ -22,77 +25,199 @@ impl FrameState {
     }
 }
 
+/// A position in the sparse memory manager containing both the region index
+/// and the frame index within that region. This ensures that region and index
+/// are always consistent.
+#[derive(Debug, Copy, Clone)]
+struct RegionFrameIndex {
+    region_idx: usize,
+    frame_idx: usize,
+}
+
 /// A physical memory manager that keeps track of the state of each frame in the
-/// system.
+/// system using a sparse representation that only tracks usable memory regions.
 pub struct PhysicalMemoryManager {
-    frames: Vec<FrameState>,
-    first_free: Option<usize>,
+    regions: Vec<MemoryRegion>,
+    first_free: Option<RegionFrameIndex>,
 }
 
 impl PhysicalMemoryManager {
+    /// Creates a new manager from usable memory regions.
+    ///
+    /// # Arguments
+    /// * `regions` - Pre-allocated vector of memory regions. Each region should already have
+    ///   frames marked as Free or Allocated based on stage1 allocations.
     #[must_use]
-    pub fn new(frames: Vec<FrameState>) -> Self {
-        let first_free = frames.iter().position(|&state| state == FrameState::Free);
-        Self { frames, first_free }
+    pub fn new(regions: Vec<MemoryRegion>) -> Self {
+        let first_free = Self::find_first_free_internal(&regions);
+        Self {
+            regions,
+            first_free,
+        }
+    }
+
+    /// Find the region and local index for a given physical address
+    fn find_frame_location(regions: &[MemoryRegion], addr: u64) -> Option<RegionFrameIndex> {
+        for (region_idx, region) in regions.iter().enumerate() {
+            if let Some(frame_idx) = region.frame_index(addr) {
+                return Some(RegionFrameIndex {
+                    region_idx,
+                    frame_idx,
+                });
+            }
+        }
+        None
+    }
+
+    /// Internal helper to find the first free frame across all regions
+    fn find_first_free_internal(regions: &[MemoryRegion]) -> Option<RegionFrameIndex> {
+        for (region_idx, region) in regions.iter().enumerate() {
+            if let Some(frame_idx) = region.frames().iter().position(|&s| s == FrameState::Free) {
+                return Some(RegionFrameIndex {
+                    region_idx,
+                    frame_idx,
+                });
+            }
+        }
+        None
+    }
+
+    /// Get the current first free frame position, updating it if necessary
+    fn first_free(&mut self) -> Option<RegionFrameIndex> {
+        if self.first_free.is_none() {
+            return None;
+        }
+
+        let ff = self.first_free?;
+
+        // Verify that the cached first_free is still valid
+        #[cfg(debug_assertions)]
+        if let Some(region) = self.regions.get(ff.region_idx) {
+            if ff.frame_idx < region.len() && region.frames()[ff.frame_idx] == FrameState::Free {
+                return Some(ff);
+            }
+        }
+
+        #[cfg(not(debug_assertions))]
+        {
+            return Some(ff);
```

Comment:
Why don't you have to update anymore? You only update in debug mode. Please explain.